contracts/atomic-swap/src/contract.rs

```diff
@@ -0,0 +1,619 @@
+
+    // We delete the swap
+    atomic_swaps(&mut deps.storage).remove(id.as_bytes());
+
+    // Send all tokens out
+    let msgs = send_native_tokens(&env.contract.address, &rcpt, swap.balance);
+    Ok(HandleResponse {
+        messages: msgs,
+        log: vec![
+            log("action", "release"),
+            log("id", id),
+            log("preimage", preimage),
+            log("to", rcpt),
+        ],
+        data: None,
+    })
+}
+
+pub fn try_refund<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    id: String,
+) -> StdResult<HandleResponse> {
+    let swap = atomic_swaps_read(&deps.storage).load(id.as_bytes())?;
+    // Anyone can try to refund, as long as the contract is expired
+    if !swap.is_expired(&env) {
+        return Err(StdError::generic_err("Atomic swap not yet expired"));
+    }
+
+    let rcpt = deps.api.human_address(&swap.source)?;
+
+    // We delete the swap
+    atomic_swaps(&mut deps.storage).remove(id.as_bytes());
+
+    let msgs = send_native_tokens(&env.contract.address, &rcpt, env.message.sent_funds);
+    Ok(HandleResponse {
+        messages: msgs,
+        log: vec![log("action", "refund"), log("id", id), log("to", rcpt)],
+        data: None,
+    })
+}
+
+fn parse_hex_32(data: &str) -> StdResult<Vec<u8>> {
+    match hex::decode(data) {
+        Ok(bin) => {
+            if bin.len() == 32 {
+                Ok(bin)
+            } else {
+                Err(StdError::generic_err("Hash must be 64 characters"))
+            }
+        }
+        Err(e) => Err(StdError::generic_err(format!(
+            "Error parsing hash: {}",
+            e.to_string()
+        ))),
+    }
+}
+
+fn send_native_tokens(from: &HumanAddr, to: &HumanAddr, amount: Vec<Coin>) -> Vec<CosmosMsg> {
+    if amount.is_empty() {
+        vec![]
+    } else {
+        vec![BankMsg::Send {
+            from_address: from.into(),
+            to_address: to.into(),
+            amount,
+        }
+        .into()]
+    }
+}
+
+pub fn query<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    msg: QueryMsg,
+) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::List {} => to_binary(&query_list(deps)?),
+        QueryMsg::Details { id } => to_binary(&query_details(deps, id)?),
+    }
+}
+
+fn query_details<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    id: String,
+) -> StdResult<DetailsResponse> {
+    let swap = atomic_swaps_read(&deps.storage).load(id.as_bytes())?;
+
+    let details = DetailsResponse {
+        id,
+        hash: swap.hash,
+        recipient: deps.api.human_address(&swap.recipient)?,
+        source: deps.api.human_address(&swap.source)?,
+        end_height: swap.end_height,
+        end_time: swap.end_time,
+        balance: swap.balance,
+    };
+    Ok(details)
+}
+
+fn query_list<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>) -> StdResult<ListResponse> {
+    Ok(ListResponse {
+        swaps: all_swap_ids(&deps.storage)?,
```

Comment:
We will want pagination on this. I like the idea to keep the logic in the `all_swap_ids` function in state, just add pagination there as well. (Maybe you calculate `start_after` and `limit` here from the user input and pass in the args that need to be applied to `range`)